docs/source/torch.compiler_dynamic_shapes.md

```diff
@@ -1,129 +1,474 @@
+such as in sparse tensors, jagged tensors, and graph neural networks.
+
+Dynamic shapes do not support dynamic rank programs, programs which input tensors
+change in dimensionality, as this is uncommon and unnecessarily complex.
+
+
+## What does it mean for a size/integer to be dynamic?
+
+Dynamic shapes allow avoiding recompilations by making certain dimensions or integers
+dynamic. For example, if a function `f(x)` is compiled with a static size, it will need
+recompilation for different sizes:
+
+```{code-cell}
+import torch
+@torch.compile(dynamic=False)
+def f(x):
+     return x* x.size()[0]
+
+f(torch.rand(10))
+f(torch.rand(20))
+f(torch.rand(30))
+f(torch.rand(40))
+```
+
+In the produced output, you can see that four graphs were generated.
+See the corresponding <a href="/_static/img/dynamic_shapes/tlparse1_dynamic_shapes_false.png" target="_blank">tlparse output</a>
+
+By making the size dynamic, the function can handle various sizes without recompilation:
+
+```{code-cell}
+import torch
+@torch.compile(dynamic=True)
+def f(x):
+     return x* x.size()[0]
+
+f(torch.rand(10))
+f(torch.rand(20))
+f(torch.rand(30))
+f(torch.rand(40))
+```
+
+With dynamic shapes enabled, only one graph is created. See the
+corresponding <a href="/_static/img/dynamic_shapes/tlparse2_dynamic_shapes_true.png" target="_blank">tlparse output</a>.
+
+While compilation time differences
+are minimal for this small example, more complex use cases would show significant
+performance improvements.
+
+(what_is_a_specialization)=
+## What is a specialization?
+
+**Specialization** refers to optimizing a computational graph for specific input shapes
+by examining shape conditions during control flow. If a branch is taken based on a
+shape condition, the graph is tailored for that condition. If a new input doesn't meet
+this condition, the system will recompile the graph.
+
+Specialization allows you to create optimized computational graphs for specific input
+shapes, which can significantly improve execution speed.
+
+```{note}
+For simplicity, this example uses `@torch.compile(dynamic=True)`. Note, that
+this option is not recommended due to it being error prone.
+For a recommended way of enabling dynamic shapes, see {ref}`enable-dynamic-behavior`
+```
+
+```{code-cell}
+import torch
+@torch.compile(dynamic=True)
+def f(x):
+    if x.size()[0] == 10:
+        return x * 10
+
+    if x.size()[0] <= 30:
+        return x*200
+
+    return x*x.size()[0]
+
+f(torch.rand(10))
+f(torch.rand(20))
+f(torch.rand(30))
+f(torch.rand(40))
+f(torch.rand(50))
+```
+
+In the code above, we specialize that the graph requires an input size of 10, in which
+case it will return `x * 10`. If the input size is less than 30, it will return `x * 200`.
+In the output, you can see that this creates three graphs.
+
+See the corresponding <a href="/_static/img/dynamic_shapes/tlparse3_specialization.png" target="_blank">tlparse output</a>
```

Comment:
I think it'd be easier to understand conceptually with the arrows pointed the other direction. eg.

func(torch.randn(10)) => graph1

func(torch.randn(20)) => graph2
func(torch.randn(30)) => graph3

etc.